--- play	2010-07-05 14:36:55 +0000
+++ play	2010-07-16 10:24:53 +0000
@@ -23,7 +23,7 @@
     play_env["basedir"] = os.path.normpath(os.path.dirname(os.path.realpath(sys.argv[0])))
 
     # ~~~~~~~~~~~~~~~~~~~~~~ What is the framework id?
-    play_env["id_file"] = os.path.join(play_env['basedir'], 'id')
+    play_env["id_file"] = os.path.join(play_env['basedir'], 'etc', 'id')
     if os.path.exists(play_env["id_file"]):
         play_env["id"] = open(play_env["id_file"]).readline().strip()
     else:

